### PR TITLE
feat: implement mvs specific functions (buildList, Upgrade, Downgrade)

### LINKED ISSUE #246: Feature: Package Version Management for kpm
# Feature Request

## Is your feature request related to a problem? 
Yes, currently KCL package manager lacks an efficient way to manage different versions of the same package. This makes it difficult for users to install specific versions, upgrade, or downgrade packages.

## Describe the feature you'd like
I would like to propose a version management feature for our package manager. This feature should allow users to easily view, install, upgrade, and downgrade different versions of packages. It should also handle dependencies version conflicts between different versions of the same package to ensure that all dependencies are met when installing, upgrading, or downgrading a package.

## Pretest

To apply for this mentorship, finish the pre-tests.
- Pre-test link: https://github.com/kcl-lang/kpm/issues/263
- Pre-test link: https://github.com/kcl-lang/kpm/issues/269

**Don't worry that you may not be able to complete all the pretests, you can choose any pretests you can do, and we will consider the final choice according to all your work. You can also try some other non-pretest work, after all, that will help you understand the whole project**

## Appendix

- kpm download the dependencies here : https://github.com/kcl-lang/kpm/blob/bd65b2a7cf109c3859bfd1bdfd315bc93a8e5caa/pkg/client/client.go#L1095

- When kpm downloads dependencies, it recursively adds all the dependencies of the dependencies to the current project. Here: https://github.com/kcl-lang/kpm/blob/bd65b2a7cf109c3859bfd1bdfd315bc93a8e5caa/pkg/client/client.go#L1146

- The Pre-Test https://github.com/kcl-lang/kpm/issues/263 can follow this approach https://github.com/kcl-lang/kpm/blob/bd65b2a7cf109c3859bfd1bdfd315bc93a8e5caa/pkg/client/client.go#L1095, and calculate the dependent graph during the download process.

- This issue can follow this approach https://github.com/kcl-lang/kpm/blob/bd65b2a7cf109c3859bfd1bdfd315bc93a8e5caa/pkg/client/client.go#L1095, when download different versions of the same package, design a selection strategy, select the appropriate version to download, and report an error if the appropriate version is not found.

- When you design a selection strategy,  you can refer to [golang mvs](https://go.dev/ref/mod#minimal-version-selection).

### PR DESCRIPTION
**Pull request by AkashKumar7902**
Implementation of previous, upgrade, required, max from go mvs

#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y 

fix #246

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [x] Contains syntax changes
- [x] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [ ] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

### CODE REVIEW

**Comment by Peefy** [comment]
@zong-zhe Can you help review and provide some more professional comments and guidance? Thank you!

File: pkg/package/package.go [hunk 1]
```diff
@@ -20,6 +22,8 @@ type KclPkg struct {
 	// The dependencies in the current kcl package are the dependencies of kcl.mod.lock,
 	// not the dependencies in kcl.mod.
 	Dependencies
+	// minimal build list for the current kcl package.
```
Comment (Peefy) [inline pkg/package/package.go:25]:
```suggestion
	// BuildList denotes the minimal build list for the current kcl package.
```

File: pkg/mvs/mvs.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	}
+	return v1
+}
+
+func (r ReqsMap) Upgrade(m module.Version) (module.Version, error) {
+	panic("unimplemented")
+}
+
+func (r ReqsMap) Previous(m module.Version) (module.Version, error) {
+	panic("unimplemented")
```
Comment (Peefy) [inline pkg/mvs/mvs.go:29]: Perhaps we can break down the unfinished features into smaller granularity PR.

File: pkg/client/test_data/add_with_git_commit/test_pkg/kcl.mod.lock.expect [hunk 3]
```diff
@@ -2,7 +2,8 @@
   [dependencies.catalog]
     name = "catalog"
     full_name = "catalog_a29e3db"
-    version = "a29e3db"
+    version = "0.1.0"
     sum = "kFmlrYJbJUFFTEXjC9cquc80WB+UpZ/6oMPKrfgyeks="
     url = "https://github.com/KusionStack/catalog.git"
     commit = "a29e3db"
+    version = "0.1.0"
```
Comment (Peefy) [inline pkg/client/test_data/add_with_git_commit/test_pkg/kcl.mod.lock.expect:9]: The version field appears twice.

File: pkg/package/package.go (hunk 1 above)
Comment (AkashKumar7902) [reply pkg/package/package.go]: done!

File: pkg/mvs/mvs.go (hunk 2 above)
Comment (AkashKumar7902) [reply pkg/mvs/mvs.go:29]: indeed ! I will add the implementation of these in the upgrade/downgrade PR

File: pkg/client/test_data/add_with_git_commit/test_pkg/kcl.mod.lock.expect (hunk 3 above)
Comment (AkashKumar7902) [reply pkg/client/test_data/add_with_git_commit/test_pkg/kcl.mod.lock.expect:8]: remove the duplicate !

**Comment by Peefy** [comment]
The PR conflict, please rebase the main branch.

**Review by Peefy — APPROVED** [review]
LGTM

**Comment by AkashKumar7902** [comment]
@Peefy I have resolved the conflict, also current lint failure is due to the code imported from go codebase, things might break if we change 3rdparty code to fix lint issue.

**Comment by AkashKumar7902** [comment]
I will send another PR containing all the cmd commands.

**Comment by Peefy** [comment]
> @Peefy I have resolved the conflict, also current lint failure is due to the code imported from go codebase, things might break if we change 3rdparty code to fix lint issue.

OK, you can fix the unit tests firstly.

File: pkg/cmd/cmd_tidy.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/urfave/cli/v2"
+	"golang.org/x/mod/module"
+	"kcl-lang.io/kpm/pkg/client"
+	"kcl-lang.io/kpm/pkg/env"
+	pkg "kcl-lang.io/kpm/pkg/package"
+	"kcl-lang.io/kpm/pkg/reporter"
+)
+
+// NewTidyCmd new a Command for `kpm graph`.
+func NewTidyCmd(kpmcli *client.KpmClient) *cli.Command {
```
Comment (Peefy) [inline pkg/cmd/cmd_tidy.go:19]: We have the update command already and no need for tidy cmmand.

File: pkg/package/modfile_test.go [hunk 5]
```diff
@@ -235,3 +235,16 @@ func TestGetFilePath(t *testing.T) {
 	assert.Equal(t, mfile.GetModFilePath(), filepath.Join(testPath, MOD_FILE))
 	assert.Equal(t, mfile.GetModLockFilePath(), filepath.Join(testPath, MOD_LOCK_FILE))
 }
+
+func TestGenSource(t *testing.T) {
```
Comment (Peefy) [inline pkg/package/modfile_test.go:239]: TestAddWithGitCommit failed on windows in CI.

Comment (AkashKumar7902) [reply pkg/package/modfile_test.go:239]: Fixed it!

**Pull request merged**